hyperspy/io_plugins/jeol.py

```diff
@@ -453,64 +594,394 @@ def parsejeol(fd):
+            # record start point of frame and skip frame
+            frame_start_index[frame_num] = p_start
+            p_start += readframe_dummy(rawdata[p_start:])
+            frame_num += 1
+            continue
+        else:
+            frame_start_index[frame_idx] = p_start  # = end of last frame
+
+        if frame_idx < frame_shifts.size:
+            fs = frame_shifts[frame_idx]
+        else:
+            fs = np.zeros(3, np.uint16)
+            _logger.info(f"Size of frame_shift array is too small. The frame {frame_idx} is not moved.")
+        length, frame_data, has_em, valid, max_valid = readframe(
+            rawdata[p_start:], 1,
+            hypermap[target_frame_num], em_image[target_frame_num],
+            width, height, channel_number,
+            width_norm, height_norm, rebin_energy,
+            fs[1], fs[0], fs[2], max_value)
+        has_em_image = has_em_image or has_em
+        if length == 0: # no data
+            break
+        if valid or not only_valid_data:
+            # accept last frame
+            if lazy:
+                data_list.append(frame_data)
+            frame_num += 1
+            target_frame_num += frame_step
+        else:
+            # incomplete data, not accepted
+            if sum_frames:
+                # subtract signal counts of last frame
+                _ = readframe(rawdata[p_start:], -1,
+                              hypermap[target_frame_num], em_image[target_frame_num],
+                              width, height, channel_number,
+                              width_norm, height_norm, rebin_energy,
+                              fs[1], fs[0],fs[2],  max_value)
+                _logger.info("The last frame (sweep) is incomplete because the acquisition stopped during this frame. The partially acquired frame is ignored. Use 'sum_frames=False, only_valid_data=False' to read all frames individually, including the last partially completed frame.")
+            break
+            # else:
+            #    pass
+
+        p_start += length
+    if not lazy:
+        if sum_frames:
+            # the first frame has integrated intensity
+            return hypermap[0,:height,:width], em_image[0,:height,:width], has_em_image, frame_num, frame_start_index, valid, max_shift, frame_shifts
+        else:
+            return hypermap[:target_frame_num,:height,:width], em_image[:target_frame_num,:height,:width], has_em_image, frame_num, frame_start_index, valid, max_shift, frame_shifts
+        
+    # for lazy loading
+
+    class DenseSliceCOO(sparse.COO):
+        """Just like sparse.COO, but returning a dense array on indexing/slicing"""
+
+        def __getitem__(self, *args, **kwargs):
+            obj = super().__getitem__(*args, **kwargs)
+            try:
+                return obj.todense()
+            except AttributeError:
+                # Indexing, unlike slicing, returns directly the content
+                return obj
```

Comment:
Instead of copying it here, it is better to import it from `hyperspy/hyperspy/misc/io/fei_stream_readers` to avoid duplicating code.
Ideally, `DenseSliceCOO` should be moved somewhere else because it is generic enough to be used elsewhere than in the velox end reader but we can leave this for another time!

Reply:
fixed.